wire-ios-sync-engine/Source/Use cases/ImportBackupUseCase/ImportBackupUseCase.swift

```diff
@@ -0,0 +1,186 @@
+    let appStateUpdater: ImportBackupAppStateUpdaterProtocol
+
+    let sharedContainerURL: URL
+    let logger: WireLogger
+
+    private let workerQueue = DispatchQueue(label: "import-backup")
+
+    func invoke(url: URL, password: String) async throws {
+
+        switch BackupFileExtensions(rawValue: url.pathExtension) {
+
+        case .fileExtensionWithUnderscore, .fileExtensionWithHyphen:
+            try await importIOSBackup(url, password)
+
+        case nil:
+            throw BackupError.invalidFileExtension
+        }
+    }
+
+    private func importIOSBackup(_ url: URL, _ password: String) async throws {
+
+        let selfClientBackup: [String: Any]
+        let account: Account
+        if let userSession = userSession() {
+
+            account = userSession.account
+
+            appStateUpdater.reportImportProgress(progress: 0.5)
+
+            let unzippedURL = try await decryptAndUnzipBackup(
+                url: url,
+                password: password,
+                accountID: userSession.account.userIdentifier
+            )
+
+            selfClientBackup = await userSession.managedObjectContext.perform {
+                userSession.selfUserClient?.backup() ?? [:]
+            }
+
+            await appStateUpdater.reportMigrationNeeded()
+
+            // user session will be destroyed now
+            _ = try await entityStorage.replacePersistentStore(
+                accountIdentifier: userSession.account.userIdentifier,
+                from: unzippedURL,
+                applicationContainer: sharedContainerURL,
+                dispatchGroup: dispatchGroup
+            )
+
+        } else {
+            throw BackupError.noActiveAccount
+        }
+
+        // import the self client
+        let temporaryStack = try await entityStorage.contextProvider(
+            account: account,
+            applicationContainer: sharedContainerURL,
+            dispatchGroup: dispatchGroup
+        )
+        try await temporaryStack.viewContext.perform {
+            let userClient = UserClient.restore(from: selfClientBackup, context: temporaryStack.viewContext)
+
+            // TODO: do we need to do this for all contexts?
+            userClient.markAsSelfClient()
+
+            try temporaryStack.viewContext.save()
+        }
+
+        await appStateUpdater.selectAccountAndTriggerSlowSync(account)
+    }
+
+    private func decryptAndUnzipBackup(url: URL, password: String, accountID: UUID) async throws -> URL {
+        logger.debug("coordinated file access at: \(url.absoluteString)")
+
+        let decryptedURL = url
+            .deletingLastPathComponent()
+            .appendingPathComponent(UUID().uuidString)
+        let unzippedURL = CoreDataStack.importsDirectory
+            .appendingPathComponent(url.deletingPathExtension().lastPathComponent)
+
+        return try await withCheckedThrowingContinuation { continuation in
+            workerQueue.async(group: dispatchGroup) {
```

Comment:
I wonder why we use the workerQueue here and the dispatchGroup, I think we could get rid of it

Reply:
I have no idea, hence I tried to copy the code without changing it.

Reply:
the workerQueue was there not to block the mainthread, now that this uses asymc/await I think we can remove it and not depend on the dispatchGroup, could you try?

Reply:
done: fc9558406539be838913b00b6c48dcf26b711a70